src/python/grpcio_reflection/grpc_reflection/v1alpha/proto_reflection_descriptor_database.py

```diff
@@ -0,0 +1,140 @@
+    ErrorResponse,
+    ExtensionNumberResponse,
+    ExtensionRequest,
+    FileDescriptorResponse,
+    ListServiceResponse,
+    ServerReflectionRequest,
+    ServerReflectionResponse,
+    ServiceResponse,
+)
+from grpc_reflection.v1alpha.reflection_pb2_grpc import ServerReflectionStub
+import logging
+
+
+class ProtoReflectionDescriptorDatabase(DescriptorDatabase):
+    """
+    ProtoReflectionDescriptorDatabase takes a stub of ServerReflection and provides the methods defined by
+    DescriptorDatabase interfaces. It can be used to feed a DescriptorPool instance.
+
+    Python implementation based on C++ version found here:
+      https://github.com/grpc/grpc/blob/v1.39.1/test/cpp/util/proto_reflection_descriptor_database.cc
+    while implementing the interface given here:
+      https://googleapis.dev/python/protobuf/3.17.0/google/protobuf/descriptor_database.html
+
+    """
+
+    def __init__(self, channel: grpc.Channel):
+        DescriptorDatabase.__init__(self)
+        self._logger = logging.getLogger(__name__)
+        self._stub = ServerReflectionStub(channel)
+        self._known_files: Set[str] = set()
+        self._cached_extension_numbers: Dict[str, List[int]] = dict()
+        self._cached_extension_files: Dict[str, FileDescriptorProto] = dict()
+
+    def _DoOneRequest(
+        self, request: ServerReflectionRequest
+    ) -> ServerReflectionResponse:
+        response = self._stub.ServerReflectionInfo(iter([request]))
+        res = next(response)
+        if res.WhichOneof("message_response") == "error_response":
+            raise grpc.RpcError(res.error_response)
+        return res
+
+    def _AddFileFromResponse(self, file_descriptor: FileDescriptorResponse):
+        protos: List[bytes] = file_descriptor.file_descriptor_proto
+        for proto in protos:
+            desc = FileDescriptorProto()
+            desc.ParseFromString(proto)
+            if desc.name not in self._known_files:
+                self._logger.info("Got new file {}".format(desc.name))
+                self._known_files.add(desc.name)
+                self.Add(desc)
+
+    def GetServices(self):
+        request = ServerReflectionRequest(list_services="")
+        response = self._DoOneRequest(request)
+        list_services: ListServiceResponse = response.list_services_response
+        services: List[ServiceResponse] = list_services.service
+        return [service.name for service in services]
+
+    def FindAllExtensionNumbers(self, extendee_name):
+        if extendee_name in self._cached_extension_numbers:
+            return self._cached_extension_numbers[extendee_name]
+        request = ServerReflectionRequest(all_extension_numbers_of_type=extendee_name)
+        try:
+            response = self._DoOneRequest(request)
+        except grpc.RpcError as e:
+            error_response: ErrorResponse = e.args[0]
+            if error_response.error_code == grpc.StatusCode.NOT_FOUND:
+                # Seems that this is "OK"
```

Comment:
Why NOT_FOUND can be considered ok?

Reply:
TBH, I don't remember, but it makes half sense given what I wrote about KeyError in the other comment thread. Anyway, I'll remove it for now.

Reply:
Well, it depends on what the expected behavior for the DescriptorDatabase should be. If a descriptor is not found, should it return an empty value or raise an error?